File: electrobun/src/views/mainview/lib/split-node.ts
import type { SplitNodeData, SplitDirection } from "../../../shared/types";

let _id = 0;
function uid(): string {
  return `sn-${++_id}-${Math.random().toString(36).slice(2, 8)}`;
}

export function createLeaf(tabId: string): SplitNodeData {
  return { type: "tab", id: uid(), tabId };
}

export function allTabIds(node: SplitNodeData): string[] {
  if (node.type === "tab") return [node.tabId];
  return [...allTabIds(node.first), ...allTabIds(node.second)];
}

export function findLeaf(
  node: SplitNodeData,
  tabId: string
): SplitNodeData | null {
  if (node.type === "tab") return node.tabId === tabId ? node : null;
  return findLeaf(node.first, tabId) ?? findLeaf(node.second, tabId);
}

/** Split the leaf containing `tabId`, placing `newTabId` beside it. Returns new tree. */
export function splitTab(
  node: SplitNodeData,
  tabId: string,
  newTabId: string,
  direction: SplitDirection
): SplitNodeData {
  if (node.type === "tab") {
    if (node.tabId === tabId) {
      return {
        type: "split",
        id: uid(),
        direction,
        first: createLeaf(tabId),
        second: createLeaf(newTabId),
        ratio: 0.5,
      };
    }
    return node;
  }

  const newFirst = splitTab(node.first, tabId, newTabId, direction);
  const newSecond = splitTab(node.second, tabId, newTabId, direction);
  if (newFirst === node.first && newSecond === node.second) return node;
  return { ...node, first: newFirst, second: newSecond };
}

/** Remove a tab from the tree, collapsing its parent split. Returns new tree or null if empty. */
export function removeTab(
  node: SplitNodeData,
  targetId: string
): SplitNodeData | null {
  if (node.type === "tab") {
    return node.tabId === targetId ? null : node;
  }

  // If one direct child is the target, return the other
  if (node.first.type === "tab" && node.first.tabId === targetId)
    return node.second;
  if (node.second.type === "tab" && node.second.tabId === targetId)
    return node.first;

  // Recurse
  const newFirst = removeTab(node.first, targetId);
  const newSecond = removeTab(node.second, targetId);

  if (newFirst === null) return newSecond;
  if (newSecond === null) return newFirst;
  if (newFirst === node.first && newSecond === node.second) return node;
  return { ...node, first: newFirst, second: newSecond };
}

/** Update the ratio of a specific split node. */
export function setRatio(
  node: SplitNodeData,
  nodeId: string,
  ratio: number
): SplitNodeData {
  const clamped = Math.max(0.1, Math.min(0.9, ratio));
  if (node.type === "tab") return node;
  if (node.id === nodeId) return { ...node, ratio: clamped };
  const newFirst = setRatio(node.first, nodeId, clamped);
  const newSecond = setRatio(node.second, nodeId, clamped);
  if (newFirst === node.first && newSecond === node.second) return node;
  return { ...node, first: newFirst, second: newSecond };
}

File: Sources/SplitNode.swift
import Foundation

/// Binary tree representing how tabs are arranged in a split layout.
/// Each leaf holds a tab ID; each branch splits two children side by side or stacked.
@MainActor
final class SplitNode: Identifiable, ObservableObject {
    let id: UUID

    enum Content {
        case tab(UUID)
        case split(direction: SplitDirection, first: SplitNode, second: SplitNode, ratio: CGFloat)
    }

    enum SplitDirection {
        case horizontal // side by side
        case vertical   // top and bottom
    }

    @Published var content: Content

    init(tabId: UUID) {
        self.id = UUID()
        self.content = .tab(tabId)
    }

    private init(direction: SplitDirection, first: SplitNode, second: SplitNode, ratio: CGFloat) {
        self.id = UUID()
        self.content = .split(direction: direction, first: first, second: second, ratio: ratio)
    }

    /// The tab ID if this is a leaf.
    var tabId: UUID? {
        if case .tab(let id) = content { return id }
        return nil
    }

    /// All tab IDs visible in this layout.
    var allTabIds: [UUID] {
        switch content {
        case .tab(let id):
            return [id]
        case .split(_, let first, let second, _):
            return first.allTabIds + second.allTabIds
        }
    }

    /// Find the leaf containing the given tab ID.
    func findLeaf(for targetId: UUID) -> SplitNode? {
        switch content {
        case .tab(let id):
            return id == targetId ? self : nil
        case .split(_, let first, let second, _):
            return first.findLeaf(for: targetId) ?? second.findLeaf(for: targetId)
        }
    }

    /// Split the leaf containing `tabId`, placing `newTabId` next to it.
    @discardableResult
    func splitTab(_ tabId: UUID, with newTabId: UUID, direction: SplitDirection) -> Bool {
        guard let leaf = findLeaf(for: tabId) else { return false }
        guard case .tab(_) = leaf.content else { return false }

        let firstChild = SplitNode(tabId: tabId)
        let secondChild = SplitNode(tabId: newTabId)

        leaf.content = .split(
            direction: direction,
            first: firstChild,
            second: secondChild,
            ratio: 0.5
        )
        return true
    }

    /// Remove a tab from the layout, collapsing its parent split.
    /// Returns true if found and removed.
    func removeTab(_ targetId: UUID) -> Bool {
        switch content {
        case .tab(_):
            return false // Can't remove self from self

        case .split(_, let first, let second, _):
            if case .tab(let id) = first.content, id == targetId {
                self.content = second.content
                return true
            }
            if case .tab(let id) = second.content, id == targetId {
                self.content = first.content
                return true
            }
            return first.removeTab(targetId) || second.removeTab(targetId)
        }
    }

    /// Update the split ratio.
    func setRatio(_ ratio: CGFloat) {
        if case .split(let dir, let first, let second, _) = content {
            content = .split(direction: dir, first: first, second: second, ratio: max(0.1, min(0.9, ratio)))
        }
    }

    // MARK: - Serialization

    final class Serialized: Codable {
        enum NodeType: String, Codable { case tab, split }
        let type: NodeType
        let tabId: String?
        let direction: String?
        let ratio: CGFloat?
        let first: Serialized?
        let second: Serialized?

        init(type: NodeType, tabId: String?, direction: String?, ratio: CGFloat?, first: Serialized?, second: Serialized?) {
            self.type = type; self.tabId = tabId; self.direction = direction
            self.ratio = ratio; self.first = first; self.second = second
        }
    }

    func serialize() -> Serialized {
        switch content {
        case .tab(let id):
            return Serialized(type: .tab, tabId: id.uuidString, direction: nil, ratio: nil, first: nil, second: nil)
        case .split(let dir, let first, let second, let ratio):
            return Serialized(
                type: .split,
                tabId: nil,
                direction: dir == .horizontal ? "horizontal" : "vertical",
                ratio: ratio,
                first: first.serialize(),
                second: second.serialize()
            )
        }
    }

    static func deserialize(_ s: Serialized) -> SplitNode? {
        switch s.type {
        case .tab:
            guard let idStr = s.tabId, let id = UUID(uuidString: idStr) else { return nil }
            return SplitNode(tabId: id)
        case .split:
            guard let dirStr = s.direction,
                  let first = s.first.flatMap({ deserialize($0) }),
                  let second = s.second.flatMap({ deserialize($0) }) else { return nil }
            let dir: SplitDirection = dirStr == "horizontal" ? .horizontal : .vertical
            let node = SplitNode(tabId: UUID()) // placeholder
            node.content = .split(direction: dir, first: first, second: second, ratio: s.ratio ?? 0.5)
            return node
        }
    }
}
